injection/sharedmain/main_test.go

```diff
@@ -0,0 +1,61 @@
+/*
+Copyright 2021 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package sharedmain
+
+import (
+	"testing"
+
+	"knative.dev/pkg/injection"
+)
+
+func TestEnabledControllers(t *testing.T) {
+	tests := []struct {
+		name                string
+		disabledControllers []string
+		ctors               []injection.NamedControllerConstructor
+		want                int
+	}{
+		{
+			name:                "zero",
+			disabledControllers: []string{"foo"},
+			ctors:               []injection.NamedControllerConstructor{{Name: "bar"}},
+			want:                1,
+		},
+		{
+			name:                "one",
+			disabledControllers: []string{"foo"},
+			ctors:               []injection.NamedControllerConstructor{{Name: "foo"}},
+			want:                0,
+		},
+		{
+			name:                "two",
+			disabledControllers: []string{"foo"},
+			ctors: []injection.NamedControllerConstructor{
+				{Name: "foo"},
+				{Name: "bar"},
+			},
+			want: 1,
```

Comment:
I think I'd probably want to have this assert that we got the expected list rather than just the size of it (seems a bit too possible for this to accidentally pass with e.g. `{"foo"}` rather than `{"bar"}`)

Reply:
OK, to do that we need to return `[]NamedControllerContructor`, meaning that we need to walk the controller constructor list twice, one for filtering named controllers (so this test can verify the list) and one for converting `[]NamedControllerContructor` to `[]ControllerContructor`.